src/exberry_int/integration_exberry.py

```diff
@@ -361,22 +366,37 @@ def mass_cancel(mass_cancel_req):
         return mass_cancel_json
 
 
-
     async def connect():
-        LOG.info(f"Connecting to the Exberry Trading API at {env.tradingApiUrl} ...")
-        ws = await ClientSession().ws_connect(env.tradingApiUrl)
-        LOG.info("...Connected to the Exberry Trading API")
+        ws = None
+        tasks = []
+        try:
+            LOG.info(f"Preparing session...")
+            session_started.clear()
+            await enqueue_outbound(make_order_book_depth(), -1)
 
-        LOG.info(f"Preparing producer coroutine...")
-        sender_task = asyncio.create_task(producer_coro(ws))
+            LOG.info(f"Connecting to the Exberry Trading API at {env.tradingApiUrl} ...")
+            ws = await ClientSession().ws_connect(env.tradingApiUrl)
+            LOG.info("...Connected to the Exberry Trading API")
 
-        LOG.info(f"Preparing consumer coroutine...")
-        receiver_task = asyncio.create_task(consumer_coro(ws))
+            LOG.info(f"Preparing producer coroutine...")
+            sender_task = asyncio.create_task(producer_coro(ws))
 
-        LOG.info(f"Requesting market session...")
-        await request_session(env.apiKey, env.secret, ws)
+            LOG.info(f"Preparing consumer coroutine...")
+            receiver_task = asyncio.create_task(consumer_coro(ws))
 
-        asyncio.gather(*[sender_task, receiver_task])
+            LOG.info(f"Requesting market session...")
+            await request_session(env.apiKey, env.secret, ws)
 
+            tasks = [sender_task, receiver_task]
+            await asyncio.gather(*tasks)
+        except Exception as e:
+            LOG.warn(f'connection error: {e}, cancelling tasks and cleaning up...')
+            for t in tasks: t.cancel()
+            if ws: await ws.close()
+
+            LOG.warn('attempting reconnect in 2 seconds...')
+            await asyncio.sleep(2)
+            await connect()
```

Comment:
This is a lot of logic to go inside an `except` block.... implication here is that it's easier for the process of handling an exception to itself throw an exception. (Which is somewhat weird.). Maybe better to factor the code such that this shutdown/retry logic is outside the `except`

Reply:
Moved this to a `finally` block.